### PR TITLE
feat:  SRS-628 - Snapshot will be created only with records which marked as public

### PR DESCRIPTION
**Pull request by midhun-aot**
- Added logic to only fetch records marked as public to create snapshot
- Added new unit test cases for the above change
- Updated RBAC for the resolver.

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Backend](https://nr-site-registry-146-backend.apps.silver.devops.gov.bc.ca)
- [Frontend](https://nr-site-registry-146-frontend.apps.silver.devops.gov.bc.ca)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-site-registry/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-site-registry/actions/workflows/merge.yml)

### CODE REVIEW

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 1]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+      }
+
+      return await this.eventsParticipantsRepo.find({
+        where: {
+          eventId: notationId,
+          srAction: SRApprovalStatusEnum.PUBLIC,
+        },
+      });
+    } catch (error) {
+      throw error;
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:170]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 2]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+    try {
+      if (siteId === '' || siteId === null) {
+        throw Error('site id cannot be empty');
+      }
+
+      return await this.eventsRepositoryRepo.find({
+        where: { siteId, srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    } catch (error) {
+      throw error;
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:187]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 3]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
     }
   }
 
+  getNotatioParticipantsForSnapshotCreation = async (notationId: string) => {
+    try {
+      if (notationId === '' || notationId === null) {
+        throw Error('notation id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:159]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 4]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+      });
+    } catch (error) {
+      throw error;
+    }
+  };
+
+  getSiteParticipantsForSnapshotCreation = async (siteId: string) => {
+    try {
+      if (siteId === '' || siteId === null) {
+        throw Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:195]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 5]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+      }
+
+      return await this.siteParticipantsRepo.find({
+        where: { siteId, srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    } catch (error) {
+      throw error;
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:203]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 6]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+    }
+  };
+
+  getSiteDocumentsForSnapshotCreation = async (siteId: string) => {
+    try {
+      if (siteId === '' || siteId === null) {
+        throw Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:211]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 7]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+      }
+
+      return await this.siteDocumentsRepo.find({
+        where: { siteId, srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    } catch (error) {
+      throw error;
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:219]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 8]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+    }
+  };
+
+  getLandHisotoriesForSnapshotCreation = async (siteId: string) => {
+    try {
+      if (siteId === '' || siteId === null) {
+        throw Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:227]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 9]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+      }
+
+      return this.landHistoriesRepo.find({
+        where: { siteId, srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    } catch (error) {
+      throw error;
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:235]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 10]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+    }
+  };
+
+  getDisclosureForSnapshotCreation = async (siteId: string) => {
+    try {
+      if (siteId === '' || siteId === null) {
+        throw Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:243]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 11]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+      }
+
+      return this.siteProfilesRepo.find({
+        where: { siteId, srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    } catch (error) {
+      throw error;
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:251]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 12]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+    }
+  };
+
+  getSubDivisionsForSnapshotCreation = async (siteId: string) => {
+    try {
+      if (siteId === '' || siteId === null) {
+        throw Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:259]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 13]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+      }
+
+      return await this.siteSubDivisionsRepo.find({
+        where: { siteId, srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    } catch (error) {
+      throw error;
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:267]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 14]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+    }
+  };
+
+  getSiteAssociationsForSnapshotCreation = async (siteId: string) => {
+    try {
+      if (siteId === '' || siteId === null) {
+        throw Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:275]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 15]
```diff
@@ -151,6 +152,121 @@ export class SnapshotsService {
...
+      }
+
+      return await this.siteAssociationsRepo.find({
+        where: { siteId, srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    } catch (error) {
+      throw error;
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:283]: Please implement HTTP exceptions with meaningful descriptive messages to enhance clarity and improve error handling.

File: backend/src/app/services/snapshot/snapshot.service.ts [hunk 16]
```diff
@@ -213,8 +327,6 @@ export class SnapshotsService {
             };
 
             snapShotsToBeSaved.push(newSnapshot);
-
-            await Promise.all(fetchEventParticipants);
           } else {
             console.log('Site id is empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts]: Please use logger service to log

File: backend/src/app/services/snapshot/snapshot.service.spec.ts [hunk 17]
```diff
@@ -787,19 +1225,210 @@ describe('SnapshotService', () => {
...
+        .spyOn(siteSubdivisionsRepository, 'find')
+        .mockResolvedValue(sampleSiteSubDivions);
+      service.getSubDivisionsForSnapshotCreation('1');
+      expect(siteSubdivisionsRepository.find).toHaveBeenCalledWith({
+        where: { siteId: '1', srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    });
+
+    it('getSubDivisionsForSnapshotCreation should throw error', async () => {
+      const mockError = new Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.spec.ts:1426]: Please implement HTTP exceptions with meaningful descriptive messages to test.

File: backend/src/app/services/snapshot/snapshot.service.spec.ts [hunk 18]
```diff
@@ -787,19 +1225,210 @@ describe('SnapshotService', () => {
...
+        .spyOn(siteAssocsRepository, 'find')
+        .mockResolvedValue(sampleSiteAssociations);
+      service.getSiteAssociationsForSnapshotCreation('1');
+      expect(siteAssocsRepository.find).toHaveBeenCalledWith({
+        where: { siteId: '1', srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    });
+
+    it('getSiteAssociationsForSnapshotCreation should throw error', async () => {
+      const mockError = new Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.spec.ts:1400]: Please implement HTTP exceptions with meaningful descriptive messages to test.

File: backend/src/app/services/snapshot/snapshot.service.spec.ts [hunk 19]
```diff
@@ -787,19 +1225,210 @@ describe('SnapshotService', () => {
...
+        .spyOn(siteProfilesRepository, 'find')
+        .mockResolvedValue(sampleProfiles);
+      service.getDisclosureForSnapshotCreation('1');
+      expect(siteProfilesRepository.find).toHaveBeenCalledWith({
+        where: { siteId: '1', srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    });
+
+    it('getDisclosureForSnapshotCreation should throw error', async () => {
+      const mockError = new Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.spec.ts:1374]: Please implement HTTP exceptions with meaningful descriptive messages to test.

File: backend/src/app/services/snapshot/snapshot.service.spec.ts [hunk 20]
```diff
@@ -787,19 +1225,210 @@ describe('SnapshotService', () => {
...
+        .spyOn(landHistoriesRepository, 'find')
+        .mockResolvedValue(sampleLandHistories);
+      service.getLandHisotoriesForSnapshotCreation('1');
+      expect(landHistoriesRepository.find).toHaveBeenCalledWith({
+        where: { siteId: '1', srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    });
+
+    it('getLandHisotoriesForSnapshotCreation should throw error', async () => {
+      const mockError = new Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.spec.ts:1348]: Please implement HTTP exceptions with meaningful descriptive messages to test.

File: backend/src/app/services/snapshot/snapshot.service.spec.ts [hunk 21]
```diff
@@ -787,19 +1225,210 @@ describe('SnapshotService', () => {
...
+        .spyOn(siteParticsRepository, 'find')
+        .mockResolvedValue(sampleSiteParticipants);
+      service.getSiteParticipantsForSnapshotCreation('1');
+      expect(siteParticsRepository.find).toHaveBeenCalledWith({
+        where: { siteId: '1', srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    });
+
+    it('getSiteParticipantsForSnapshotCreation should throw error', async () => {
+      const mockError = new Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.spec.ts:1322]: Please implement HTTP exceptions with meaningful descriptive messages to test.

File: backend/src/app/services/snapshot/snapshot.service.spec.ts [hunk 22]
```diff
@@ -787,19 +1225,210 @@ describe('SnapshotService', () => {
...
+    it('getSiteDocumentsForSnapshotCreation should be called with SRAction equals public', async () => {
+      jest.spyOn(siteDocsRepository, 'find').mockResolvedValue(sampleDocuments);
+      service.getSiteDocumentsForSnapshotCreation('1');
+      expect(siteDocsRepository.find).toHaveBeenCalledWith({
+        where: { siteId: '1', srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    });
+
+    it('getSiteDocumentsForSnapshotCreation should throw error', async () => {
+      const mockError = new Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.spec.ts:1297]: Please implement HTTP exceptions with meaningful descriptive messages to test.

File: backend/src/app/services/snapshot/snapshot.service.spec.ts [hunk 23]
```diff
@@ -787,19 +1225,210 @@ describe('SnapshotService', () => {
...
+        .spyOn(eventParticsRepository, 'find')
+        .mockResolvedValue(sampleNotationParticipants);
+      service.getNotatioParticipantsForSnapshotCreation('1');
+      expect(eventParticsRepository.find).toHaveBeenCalledWith({
+        where: { eventId: '1', srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    });
+
+    it('getNotatioParticipantsForSnapshotCreation should throw error', async () => {
+      const mockError = new Error('notation id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.spec.ts:1276]: Please implement HTTP exceptions with meaningful descriptive messages to test.

File: backend/src/app/services/snapshot/snapshot.service.spec.ts [hunk 24]
```diff
@@ -787,19 +1225,210 @@ describe('SnapshotService', () => {
...
-      expect(result).toEqual(expectedBannerType);
+      service.getNotationsForSnapshotCreation('1');
+
+      expect(eventsRepository.find).toHaveBeenCalledWith({
+        where: { siteId: '1', srAction: SRApprovalStatusEnum.PUBLIC },
+      });
+    });
+
+    it('getNotationsForSnapshotCreation should throw error', async () => {
+      const mockError = new Error('site id cannot be empty');
```
Comment (jaspalsingh-aot) [inline backend/src/app/services/snapshot/snapshot.service.spec.ts:1251]: Please implement HTTP exceptions with meaningful descriptive messages to test.

**Review by jaspalsingh-aot — CHANGES_REQUESTED** [review]

File: backend/src/app/services/snapshot/snapshot.service.ts (hunk 3 above)
Comment (midhun-aot) [reply backend/src/app/services/snapshot/snapshot.service.ts:159]: This is independent method  and no need for HTTP exception from here.

File: backend/src/app/services/snapshot/snapshot.service.ts (hunk 1 above)
Comment (midhun-aot) [reply backend/src/app/services/snapshot/snapshot.service.ts:170]: This is independent method and no need for HTTP exception from here

File: backend/src/app/services/snapshot/snapshot.service.ts (hunk 2 above)
Comment (midhun-aot) [reply backend/src/app/services/snapshot/snapshot.service.ts:187]: This is independent method and no need for HTTP exception from here

File: backend/src/app/services/snapshot/snapshot.service.ts (hunk 4 above)
Comment (midhun-aot) [reply backend/src/app/services/snapshot/snapshot.service.ts:195]: This is independent method and no need for HTTP exception from here

File: backend/src/app/services/snapshot/snapshot.service.ts (hunk 5 above)
Comment (midhun-aot) [reply backend/src/app/services/snapshot/snapshot.service.ts:203]: This is independent method and no need for HTTP exception from here

File: backend/src/app/services/snapshot/snapshot.service.ts (hunk 1 above)
Comment (nikhila-aot) [inline backend/src/app/services/snapshot/snapshot.service.ts:170]: implement logging in all the methods in this file  - start, end and catch block of the method. Also refer to the HttpExecption implemented in the other service layers and check the feasibility of having it.

Comment (midhun-aot) [reply backend/src/app/services/snapshot/snapshot.service.ts:170]: Added log statements to all the method. As discussed HTTP exception is not required here as its already handled in main service method with proper response code.

**Review by jaspalsingh-aot — APPROVED** [review]

**Pull request merged**